### PR TITLE
Fix incorrect distribution of randperm with device mps

### LINKED ISSUE #104170: torch.randperm is broken with device "mps"
### 🐛 Describe the bug

Running torch.randperm with the device as "mps" does not produce similarly distributed results to that of the cpu version.

I did not compile any statistical analysis but I think the problem can be seen quite clearly from running the command a few times.

Here is an example:
```
Python 3.11.3 (main, May 15 2023, 18:01:31) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.randperm(100, device="mps")
tensor([ 3,  4,  7, 10, 11, 13, 14, 15, 16, 17, 20, 23, 24, 25, 26, 28, 29, 30,
        31, 32, 34, 35, 36, 38, 40, 43, 44, 46, 47, 48, 49, 50, 51, 52, 53, 55,
        56, 59, 61, 62, 64, 68, 71, 74, 79, 80, 82, 83, 84, 85, 86, 88, 89, 90,
        91, 92, 95, 96, 97, 99,  0,  1,  2,  5,  6,  8,  9, 12, 18, 19, 21, 22,
        27, 33, 37, 39, 41, 42, 45, 54, 57, 58, 60, 63, 65, 66, 67, 69, 70, 72,
        73, 75, 76, 77, 78, 81, 87, 93, 94, 98], device='mps:0')
>>> torch.randperm(100, device="mps")
tensor([ 1,  5,  7,  9, 12, 13, 16, 18, 19, 21, 24, 27, 29, 31, 32, 36, 39, 42,
        43, 44, 46, 47, 50, 51, 52, 55, 61, 65, 66, 70, 72, 75, 78, 80, 81, 82,
        85, 86, 88, 89, 90, 95, 98,  0,  2,  3,  4,  6,  8, 10, 11, 14, 15, 17,
        20, 22, 23, 25, 26, 28, 30, 33, 34, 35, 37, 38, 40, 41, 45, 48, 49, 53,
        54, 56, 57, 58, 59, 60, 62, 63, 64, 67, 68, 69, 71, 73, 74, 76, 77, 79,
        83, 84, 87, 91, 92, 93, 94, 96, 97, 99], device='mps:0')
>>> torch.randperm(100, device="mps")
tensor([ 0,  1,  3,  4,  5,  7,  8, 10, 13, 15, 17, 20, 21, 23, 24, 25, 26, 27,
        30, 31, 32, 33, 34, 36, 37, 39, 43, 45, 47, 49, 51, 52, 53, 55, 56, 58,
        59, 60, 62, 66, 67, 74, 77, 80, 81, 84, 85, 87, 88, 90, 93, 97,  2,  6,
         9, 11, 12, 14, 16, 18, 19, 22, 28, 29, 35, 38, 40, 41, 42, 44, 46, 48,
        50, 54, 57, 61, 63, 64, 65, 68, 69, 70, 71, 72, 73, 75, 76, 78, 79, 82,
        83, 86, 89, 91, 92, 94, 95, 96, 98, 99], device='mps:0')
>>> torch.randperm(100, device="cpu")
tensor([38, 58, 16, 34, 51, 72, 57, 71, 94, 67, 40, 84, 69, 83,  4, 41, 70,  1,
        65, 73, 45, 92, 88, 11,  5, 48, 33,  3, 82, 22, 75, 56, 14, 95, 80, 96,
        53, 64, 60, 81, 49, 93, 20, 19, 89, 27, 13, 37, 21, 98, 62, 31, 91, 97,
        86, 32, 24, 47, 85, 77,  7, 90, 35, 28, 44, 63, 74, 46, 54, 36, 42, 66,
        43, 12,  6, 30,  8, 68, 79, 78,  0,  2, 50, 26, 55, 17, 23, 52, 15, 59,
        25, 61, 76, 10,  9, 29, 39, 99, 87, 18])
>>> torch.randperm(100, device="cpu")
tensor([24, 61, 59, 81, 90, 92, 36, 82, 37, 75, 55, 21, 77, 41, 34, 47, 45, 17,
         9, 52, 69, 71, 23, 79, 46, 50,  8, 57, 18,  7, 98,  1, 95, 31, 83, 40,
        16, 51, 86,  0, 10, 93, 67, 91, 48, 14, 22, 28, 30, 99,  2, 64, 27, 68,
        62, 29, 49, 74, 13, 73, 66, 78, 97, 25, 72, 33, 88, 38, 26, 54, 65, 58,
        84,  6, 60, 87, 63, 80, 53, 44, 15, 70, 89,  5, 42, 56, 19, 11, 94, 12,
         3, 96, 32, 76, 85,  4, 39, 35, 20, 43])
>>> torch.randperm(100, device="cpu")
tensor([85,  2, 66, 50, 40, 52, 65, 29, 12, 46, 82, 83, 48, 78, 70, 67,  9, 41,
        86, 11, 34, 77, 35, 42, 54, 91, 93, 84, 76, 31, 10, 19, 89, 68, 57, 27,
        18, 15, 13, 87, 45,  7, 36, 63, 64, 99, 24, 51,  0,  6, 73, 81, 58, 16,
        60, 92, 90, 17, 62, 39, 37, 61, 97, 74, 47, 43, 94, 59, 28,  5, 23, 95,
        33, 55, 69, 30,  4,  3, 21, 49, 20, 79, 56, 96, 38, 22, 98, 44, 25, 14,
        80, 72, 32, 71,  8, 26, 75, 88, 53,  1])
```

In the mps version seems like the indexes are separated into two increasing(?) partitions which are then concatenated together. Or even if that is incorrect at least they do not seem to be uniformly distributed.

I looked into the [pr that introduced this code](https://github.com/pytorch/pytorch/pull/91708/files) and from what I can see it basically does an argsort on a uniformly random tensor. Interestingly enough `torch.argsort(torch.rand(100, device="mps"), dim = 0)` does not seem to have this problem:

```
>>> torch.argsort(torch.rand(100, device="mps"), dim = 0)
tensor([86, 75, 89, 55, 52, 87, 85, 99,  0, 22, 70, 13, 81, 42, 69, 78, 92, 57,
        24, 44, 83, 63, 66, 98, 72, 41, 51, 56, 60, 16, 97, 39, 50, 15, 67, 45,
        30, 19, 47, 10, 79, 14, 21, 68, 20, 46, 37, 82, 96, 80, 40, 61, 38, 77,
         9, 17, 48, 49, 43,  5, 54, 27,  8, 34, 26, 25,  6, 90, 11, 84, 32, 59,
        95, 29, 36, 12, 93, 94,  1, 74, 35, 65, 28, 91, 64, 76,  3, 71, 58, 23,
        18, 31,  2, 62, 33, 53, 88, 73,  7,  4], device='mps:0')
>>> torch.argsort(torch.rand(100, device="mps"), dim = 0)
tensor([38, 21, 97, 71, 44, 18, 65, 16, 54, 70,  3, 15, 69, 12, 79, 90, 34, 88,
        92, 27, 13, 22, 59, 75, 80, 68, 63, 98, 19, 84, 96, 11, 14, 32, 48, 55,
         8, 73,  1, 58, 30, 36, 10, 93, 78, 42, 99, 45, 53, 74, 89, 20, 57, 35,
        76,  5, 81, 62, 50,  2, 31, 94, 66, 77, 29, 26, 72, 51,  4, 64,  9, 40,
        95,  7, 87, 33, 39, 61, 24, 86, 85, 37, 60, 91, 25, 82,  0, 23,  6, 46,
        52, 49, 56, 28, 17, 83, 67, 47, 43, 41], device='mps:0')
>>> torch.argsort(torch.rand(100, device="mps"), dim = 0)
tensor([28, 62, 37, 70, 97, 40, 79,  0, 27, 19, 99, 90, 55, 51, 14, 65,  6, 47,
        81, 53, 36, 86, 29, 16, 18, 33,  5,  7, 17,  2, 52, 73, 88, 23, 76, 72,
         9, 30, 50, 12, 78, 83, 57, 46, 39, 10, 42, 24, 91, 77, 49, 13, 20, 44,
        54, 87, 11, 56, 15, 75, 80, 69,  4, 25, 84, 58, 66, 68, 22,  1, 64, 60,
        61, 71, 38, 96, 48, 85, 67, 74, 93, 98, 35,  8, 32, 34, 31, 41, 43, 59,
        92, 21,  3, 63, 26, 94, 82, 95, 89, 45], device='mps:0')
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0a0+git925f0a0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.26.4
Libc version: N/A

Python version: 3.11.3 (main, May 15 2023, 18:01:31) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.25.0
[pip3] torch==2.1.0a0+git925f0a0
[conda] numpy                     1.25.0                   pypi_0    pypi
[conda] torch                     2.1.0a0+git925f0a0           dev_0    <develop>
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by Mr4k**
Fixes #104170

As noted in the above issue it seems that the code for randperm basically boils down to:
`torch.argsort(torch.rand(size, device="mps"), dim = 0)`

However it seems like in the fused(?) pytorch version the type of tensor we were drawing `torch.rand(size, device="mps")` from was int64 with an inclusive(?) upper bound of 1. This caused everything to be sorted into two groups (if you drew 0 or 1) each monotonically ascending due to sort tie breaking.

One way to fix this is to  just generate the random tensor as float64s with an upper bound of 1.0 instead of int64s. An alternative to to just set the upper bound to max int 64. 

~I choose the float64 one basically on a coin flip b/c I couldn't tell the original contributor's intent (due to mixed up upper bounds and type) but would be happy to change to use int64 and max int 64 as an upper bound instead if that's better.~

Edit on second thought I don't like using floats from 0.0 to 1.0 as there are fewer of them in that range than int64s from 0 to int 64 max_value. I also suspect integer math might be faster but need to benchmark this tomorrow.

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Distributions.mm [hunk 1]
```diff
@@ -412,8 +412,8 @@ Tensor normal_mps(const Tensor& mean, const Tensor& std, c10::optional<Generator
   };
 
   return mps::random_mps_impl<int64_t>(result,
-                                       0.0,
-                                       1.0,
+                                       std::numeric_limits<int64_t>::min(),
```
Comment (Mr4k) [inline aten/src/ATen/native/mps/operations/Distributions.mm:415]: Changed this lower bound from 0 to -int64 min b/c more numbers -> few collisions which need to be tie broken which I think leads to a better distribution. Not sure if there are speed or portability problems here though. I will run benchmarks tomorrow.

File: aten/src/ATen/native/mps/operations/Distributions.mm [hunk 2]
```diff
@@ -412,8 +412,8 @@ Tensor normal_mps(const Tensor& mean, const Tensor& std, c10::optional<Generator
...
+                                       std::numeric_limits<int64_t>::max(),
```
Comment (Mr4k) [inline aten/src/ATen/native/mps/operations/Distributions.mm:416]: Changed this upper bound from 1 to int64 max which solves the original problem by making the range wider. Used int64 max again b/c more numbers -> fewer collisions. I will run benchmarks tomorrow.

**Comment by kit1980** [comment]
@Mr4k Thanks for you PR. Please sign the CLA.

**Comment by Mr4k** [comment]
@kit1980 sorry about that! Should be signed now!

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
Thank you for your change. In the long run, it would be good to compare it more thoroughly to CPU/CUDA implementation. I.e. binning it to 2 buckets seems wrong, but I wonder if it would result in better distribution, if one to range from 0 to `n` instead of `-2**31` to `+2**31`

**Comment by Mr4k** [comment]
@malfet thanks for reviewing!

Re: the number of buckets:
I think generally the more buckets the better assuming we want a uniform distribution over all possible permutations.

Here is a simple example:
Let's say we have an array of size 2. There are two possible arrangements for the array's indexes:
```
[0, 1]
[1, 0]
```

If we are drawing the permutations from a uniform distribution we would want the probability of each permutation to be equal:

```
p([0, 1]) = 0.5
p([1, 0]) = 0.5
```

Currently the way this method works is to assign each index a random int64 value and sort them based on that value. For example:

```
0 -> 7
1 -> 4
```
In this case we assign 0 to 7 and 1 to 1. Sorting the indexes based on their assigned values (in increasing order) gives us the permutation `[1, 0]`.

Since the values are drawn without replacement occasionally both indexes might be assigned the same value. In that case ties are broken based on the value of the indexes themselves. If 1 and 0 are both assigned 10 for example the resulting order will always be `[0, 1]`

So now we can compute the probability of ending up with each of the two permutations of our size 2 array based on the number of buckets :

Let `v(i)` denote the random value assigned to index i.

```
p([0, 1]) = p(v(0) < v(1)) + p(v(0) = v(1))
p([1, 0]) = p(v(0) > v(1))
```

Because of symmetry:
```
p(v(0) < v(1)) = p(v(0) > v(1))
```

so the difference between `p([0, 1])` and `p([1, 0])` is `p(v(0) = v(1))` which is the probability of a collision. Since we want `p([0, 1])` and `p([1, 0])` to be as close to as equal as possible we want to minimize `p(v(0) = v(1))`. The more buckets the smaller that value will be. So more buckets will always be better.

The cuda kernel uses a [similar approach](https://github.com/pytorch/pytorch/blob/b76a040b18e2e9f768f566ef6abbc4b5cb0ab48f/aten/src/ATen/native/cuda/Randperm.cu#L93)

**Comment by Mr4k** [comment]
@malfet sorry to tag you again I just did a quick edit in the post above to reference an extra step the cuda kernel uses to make sure their distribution is correctly uniform.
